cider/src/debugger/source/structures.rs

```diff
@@ -26,14 +28,67 @@ pub struct GroupContents {
     pub start_line: u64,
     pub end_line: u64,
 }
+/// first item is group name and second is component name
+type GroupName = (String, String);
 
 /// impl struct with path and number
 #[derive(Debug, Clone)]
 /// NewSourceMap contains the group name as the key and the line it lies on with
 ///  as respect to its corresponding .futil file
-pub struct NewSourceMap(HashMap<(String, String), GroupContents>);
+pub struct NewSourceMap(HashMap<GroupName, GroupContents>);
 
 impl NewSourceMap {
+    pub fn generate_from_source_info(
+        table: &SourceInfoTable,
+        ctx: &Context,
+    ) -> CiderResult<NewSourceMap> {
+        let mut map: HashMap<GroupName, GroupContents> = HashMap::new();
+        // iterate over groups
+        ctx.primary.groups.iter().for_each(|(g_idx, group)| {
```

Comment:
style nit: I generally prefer using a direct `for` loop over `for_each` since
the latter tends to be easier to understand